File: 0206-reverse-linked-list/0206-reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode* prev = NULL; // This will be the new tail of the list, so it initially points to NULL.
    struct ListNode* curr = head; // Current node we're processing.

    while (curr != NULL) {
        struct ListNode* nextTemp = curr->next; // Temporarily store the next node.
        curr->next = prev;                      // Reverse the current node's pointer.
        prev = curr;                            // Move 'prev' one step forward (to the current node).
        curr = nextTemp;                        // Move 'curr' one step forward (to the next node).
    }

    return prev; // At the end, 'prev' will be the new head of the reversed list.
}

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Initialize variables
        prev = None  # This will become the new head of the reversed list
        curr = head  # Start with the current node as the head of the list

        while curr:
            next_node = curr.next  # Store the next node
            curr.next = prev       # Reverse the current node's pointer
            prev = curr            # Move the 'prev' pointer to the current node
            curr = next_node       # Move to the next node in the list

        # At the end, 'prev' will be the new head of the reversed list
        return prev
